### PR TITLE
[kernel] Fix #13495: incompleteness in cases typing for cumulative inductive types

### LINKED ISSUE #13495: Failure of subject reduction due to incompleteness in the type-checker
#### Description of the problem

The type-checker is incomplete in presence of cumulative inductive types, 
as realized while formalizing the completeness proof of PCUIC in MetaCoq 
(problem found by @MevenBertrand).

```
Universe u.
(* Constraint Set < u. *)
Polymorphic Cumulative Record pack@{u} := Pack { pack_type : Type@{u} }.
(* u is covariant *)

Polymorphic Definition pack_id@{u} (p : pack@{u}) : pack@{u} := 
  match p with
  | Pack T => Pack T
  end.
Definition packid_nat (p : pack@{Set}) := pack_id@{u} p.
(* No constraints: Set <= u *)

Fail Definition sr_breaks := Eval compute in packid_nat.
(* Incorrect elimination of "p" in the inductive type "pack":
  ill-formed elimination predicate. 
  
  This is because it tries to enforce Set = u
  *)
```

#### Coq Version

Coq 8.11, but all other versions with cumulative inductives are affected.
The problem lies in typeops.ml and inductive.ml#is_correct_arity. The arity of the predicate is checked to be correct using conversion while cumulativity is needed. A similar problem happens for branches.

### PR DESCRIPTION
**Pull request by mattam82**
**Kind:** bug fix

Fixes / closes #13495

- [x] Added / updated test-suite

- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

File: kernel/inductive.ml [hunk 1]
```diff
@@ -330,33 +330,44 @@ let check_allowed_sort ksort specif =
...
+        try conv_leq env dep_ind a1'
+        with NotConvertible -> raise (LocalArity None)
+      in
+      let () = check_allowed_sort ksort specif in
+      (* We return the "higher" inductive universe instance from the predicate,
+         the branches must be typeable using these universes. *)
+      let (pind, _args) =
+        try find_rectype env a1'
+        with Not_found ->
+          anomaly (Pp.str "Not an inductive type but convertible to an inductive type")
```
Comment (SkySkimmer) [inline kernel/inductive.ml:363]: I don't think we need to catch here

Comment (mattam82) [reply kernel/inductive.ml:363]: Well, I didn't know which was preferable, we know that a1' must be an inductive given the previous cumulativity check.

Comment (mattam82) [reply kernel/inductive.ml]: I removed it but added an explanatory comment.

**Comment by mattam82** [comment]
Is is really expected that in bedrock2 we have:
--------------------
 509m17.93s | 1336072 ko | Kami/Ex/Multiplier64.vo                                                                        
 477m26.73s | 1130888 ko | Kami/Ex/Multiplier32.vo      
cc: @JasonGross ?

**Comment by JasonGross** [comment]
This is a question for @samuelgruetter @andres-erbsen

**Comment by mattam82** [comment]
@SkySkimmer looks ready to me

**Comment by gares** [comment]
I'm running the bench at https://gitlab.com/coq/coq/-/jobs/877642080 since I plan to backport it and I've already run benchemarks for 8.13

**Comment by samuelgruetter** [comment]
> Is it really expected that in bedrock2 we have:
> 
> 509m17.93s | 1336072 ko | Kami/Ex/Multiplier64.vo
> 477m26.73s | 1130888 ko | Kami/Ex/Multiplier32.vo

I didn't write these files, but just checked on our Travis CI, where they take a bit more than half a minute and a bit more than one minute, respectively, both on [8.12.0](https://travis-ci.com/github/mit-plv/bedrock2/jobs/448334883#L2754-L2757) and [master](https://travis-ci.com/github/mit-plv/bedrock2/jobs/448334881#L3425-L3428)

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot: merge now

**Comment by SkySkimmer** [comment]
~~~
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                             │                         │                                       │                                       │                         │                 │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD  PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                   coq-color │  512.52   513.63  -0.22 │  1430319274590   1429729756532   0.04 │  1795642748586   1795366857153   0.02 │ 1077576  1077656  -0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-fiat-parsers │  637.55   638.36  -0.13 │  1785077605236   1786780965960  -0.10 │  2728888254238   2728457877934   0.02 │ 3639504  3639580  -0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-fourcolor │ 2583.12  2584.42  -0.05 │  7200763658423   7205347363884  -0.06 │ 12433717555969  12433613844404   0.00 │  722392   722392   0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-math-classes │  212.74   212.78  -0.02 │   594328071167    592716338966   0.27 │   813134821553    813051098619   0.01 │  506228   506268  -0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-bedrock2 │  227.52   227.46   0.03 │   630319869040    628515893429   0.29 │  1043195017226   1043014790573   0.02 │ 1085136  1085224  -0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-rewriter │  723.98   723.68   0.04 │  2009277366882   2009873225614  -0.03 │  3105250642356   3104239757112   0.03 │  988936   989296  -0.04 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                   coq-flocq │  156.55   156.42   0.08 │   435101783909    434691776969   0.09 │   541172184342    541171927673   0.00 │  820832   820784   0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-coqprime │   79.39    79.31   0.10 │   217980391564    217967063639   0.01 │   306121606853    306110564507   0.00 │  761676   761732  -0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│             coq-fiat-crypto │ 5122.28  5116.25   0.12 │ 14216993733403  14209555901223   0.05 │ 24815848143185  24814043797988   0.01 │ 2281560  2267400   0.62 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                         coq │  522.66   522.02   0.12 │  1461852732319   1464317487219  -0.17 │  1959669311567   1959772501413  -0.01 │  562192   562100   0.02 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-fiat-core │  103.00   102.83   0.17 │   293315718277    293286774796   0.01 │   390088079301    389852073964   0.06 │  479896   479760   0.03 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│ coq-rewriter-perf-SuperFast │  793.00   791.63   0.17 │  2207386935767   2205055906866   0.11 │  3476514721458   3476438011275   0.00 │  981064   981316  -0.03 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-mathcomp-ssreflect │   51.83    51.74   0.17 │   141169285961    140627177131   0.39 │   178653127647    178637170990   0.01 │  513816   513800   0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-compcert │  642.94   641.63   0.20 │  1791795724776   1787000938203   0.27 │  2479502730755   2479057952263   0.02 │ 1046856  1046864  -0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-solvable │  171.13   170.72   0.24 │   476404920635    474740514146   0.35 │   672008506362    672045997872  -0.01 │  672756   683172  -1.52 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-engine-bench-lite │  366.79   365.82   0.27 │  1031752393870   1029489801904   0.22 │  1753463954546   1747051335772   0.37 │ 3499968  3499940   0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-perennial │ 5182.98  5163.96   0.37 │ 14391708489727  14351251129248   0.28 │ 21458127156468  21468163755049  -0.05 │ 3104404  3211872  -3.35 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-bignums │   56.96    56.74   0.39 │   157095458196    156725053898   0.24 │   211493194383    211458707704   0.02 │  476972   476868   0.02 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                    coq-corn │ 1632.23  1625.55   0.41 │  4542144511305   4524200003793   0.40 │  6946438386270   6944754291151   0.02 │  862144   860884   0.15 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                  coq-geocoq │ 1454.08  1447.76   0.44 │  4050386055291   4032142423234   0.45 │  6141079667267   6141215365677  -0.00 │ 1014776  1017512  -0.27 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│             coq-lambda-rust │ 1254.57  1248.99   0.45 │  3491949088077   3474777814637   0.49 │  4984848932604   4987715356382  -0.06 │ 1095100  1117308  -1.99 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                    coq-hott │  325.32   323.84   0.46 │   883938543006    881359274772   0.29 │  1434548640534   1434429419592   0.01 │  728568   710996   2.47 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-verdi-raft │ 1227.16  1221.34   0.48 │  3416904957013   3400609970572   0.48 │  4837334056480   4837289078352   0.00 │  901732   902052  -0.04 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│          coq-mathcomp-field │  224.80   223.70   0.49 │   624779546663    622778011796   0.32 │   950949240565    951007679318  -0.01 │  672100   672244  -0.02 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-fingroup │   47.68    47.44   0.51 │   132031405643    131448367197   0.44 │   181506695334    181539454011  -0.02 │  480368   481640  -0.26 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-unimath │ 3945.59  3925.62   0.51 │ 10980340639792  10922291464814   0.53 │ 20982960823007  20983296857357  -0.00 │ 1142344  1142264   0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-coqutil │   60.81    60.44   0.61 │   168226794741    167709102579   0.31 │   209344434722    209299059413   0.02 │  524320   523820   0.10 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│        coq-mathcomp-algebra │  138.38   137.53   0.62 │   385048955077    382717753721   0.61 │   495362725874    495407218882  -0.01 │  546208   546276  -0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-mathcomp-character │  155.60   154.61   0.64 │   432604135624    430627413560   0.46 │   599583163868    599605827017  -0.00 │  708136   708404  -0.04 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-coquelicot │   68.36    67.85   0.75 │   187119100120    185944662478   0.63 │   237643465067    237552358784   0.04 │  565928   565944  -0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-mathcomp-odd-order │ 1064.10  1053.74   0.98 │  2961752547707   2932579588353   0.99 │  4692894996930   4692893684709   0.00 │  975628   975584   0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                   coq-verdi │  103.47   102.28   1.16 │   284172519346    282488038450   0.60 │   398593277526    398442690725   0.04 │  548216   546536   0.31 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘
~~~

**Comment by SkySkimmer** [comment]
I didn't notice the bench before merging, I guess it's a bit slower because we reduce `a1'` twice (to convert and to get the head out)

**Comment by mattam82** [comment]
Yes, it might be a little bit worse in general, but most of the time it will be immediate to get back the inductive (whd_all is optimized to not reduce terms headed by tInd, and the pretyper always produces an explicit inductive type for the binder I assume).

**Comment by samuelgruetter** [comment]
> 509m17.93s | 1336072 ko | Kami/Ex/Multiplier64.vo
> 477m26.73s | 1130888 ko | Kami/Ex/Multiplier32.vo

Now that this is merged, the bedrock2 CI for master times out. It didn't show up in the bench because the bench only builds a the program logic sudirectory (`bedrock2/bedrock2`).
The problem is that in `bedrock2/deps/kami/Kami/Ex/Multiplier32.v`, the `Qed` of `boothMultiplierInv_ok'`, which ran reasonably fast before, now takes hours.
Is it expected that this PR can cause such drastic Qed time changes?

**Comment by mattam82** [comment]
The one change that I think could have impact is that conversion is tested in the other direction now. Let's see if this changes anything.

**Pull request merged**